Prevent Kubelet from incorrectly interpreting "not yet started" pods as "ready to terminate pods" by unifying responsibility for pod lifecycle into pod worker

smarterclayton opened the pull request:
A number of race conditions exist when pods are terminated early in their lifecycle because components in the kubelet need to know "no running containers" or "containers can't be started from now on" but were relying on outdated state (they don't know whether the pod is setting up, tearing down, or can never be started again).

Only the pod worker knows whether containers are being started for a given pod, which is required to know when a pod is "terminated" (no running containers, none coming). Move that responsibility and podKiller function into the pod workers, and have everything that was killing the pod go into the UpdatePod loop. Have pod workers remain running until the pod is ready to be deleted in etcd (PodResourcesAreReclaimed would return true).  Split syncPod into three phases - setup, terminate containers, and cleanup pod - and have transitions between those methods be visible to other components. After this change, to kill a pod you tell the pod worker to UpdatePod({UpdateType: SyncPodKill, Pod: pod}).

Several places in the kubelet were incorrect about whether they were handling terminating (should stop running, might have
containers) or terminated (no running containers) pods. The pod worker exposes methods that allow other loops to know when to set up or tear down resources based on the state of the pod - these methods remove the possibility of race conditions by ensuring a single component is responsible for knowing each pod's allowed state and other components
simply delegate to checking whether they are in the window by UID.

In addition, removing containers now no longer blocks final pod deletion in the API server and are handled as background cleanup.

See https://docs.google.com/document/d/1DvAmqp9CV8i4_zYvNdDWZW-V0FMk1NbZ8BAOvLRD3Ds/edit# for details

TODO:

* [x] Implement context cancellation on syncPod and terminatingPod when necessary
* [x] Add an e2e test like "pod submit and delete" with init containers https://github.com/kubernetes/kubernetes/pull/103128
* [ ] Add an e2e test that verifies that logs remain and are accessible until the pod object is deleted (set a 15s grace period, write a log message every second)
* [x] Thoroughly review that no regressions in cleanup are present - setup in loops should use the right conditions, teardown should also use them
* [x] Improve pod sync latency by having pod worker track the latency

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind flake

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix a number of race conditions in the kubelet when pods are starting up or shutting down that might cause pods to take a long time to shut down.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

smarterclayton committed:
Keep pod worker running until pod is truly complete

A number of race conditions exist when pods are terminated early in
their lifecycle because components in the kubelet need to know "no
running containers" or "containers can't be started from now on" but
were relying on outdated state.

Only the pod worker knows whether containers are being started for
a given pod, which is required to know when a pod is "terminated"
(no running containers, none coming). Move that responsibility and
podKiller function into the pod workers, and have everything that
was killing the pod go into the UpdatePod loop. Split syncPod into
three phases - setup, terminate containers, and cleanup pod - and
have transitions between those methods be visible to other
components. After this change, to kill a pod you tell the pod worker
to UpdatePod({UpdateType: SyncPodKill, Pod: pod}).

Several places in the kubelet were incorrect about whether they
were handling terminating (should stop running, might have
containers) or terminated (no running containers) pods. The pod worker
exposes methods that allow other loops to know when to set up or tear
down resources based on the state of the pod - these methods remove
the possibility of race conditions by ensuring a single component is
responsible for knowing each pod's allowed state and other components
simply delegate to checking whether they are in the window by UID.

Removing containers now no longer blocks final pod deletion in the
API server and are handled as background cleanup. Node shutdown
no longer marks pods as failed as they can be restarted in the
next step.

See https://docs.google.com/document/d/1Pic5TPntdJnYfIpBeZndDelM-AbS4FN9H2GTLFhoJ04/edit# for details

smarterclayton commented:
/priority critical-urgent

Race conditions in the kubelet mean that quickly deleted pods can leave dangling resources for significant amounts of time (although this is enough of a change that we shouldn't rush to merge it until we're confident it's both correct, safe, and improves testing overall)

ehashman commented:
/triage accepted

msau42 commented:
/assign  @gnufied @jingxu97
for volumemanager implications

If I understand correctly, this may help resolve some of the races we've seen such as https://github.com/kubernetes/kubernetes/issues/69831, https://github.com/kubernetes/kubernetes/issues/96759, https://github.com/kubernetes/kubernetes/issues/101911

ehashman commented:
Doc is https://docs.google.com/document/d/1DvAmqp9CV8i4_zYvNdDWZW-V0FMk1NbZ8BAOvLRD3Ds/edit# (link in first comment is the RH internal draft)

smarterclayton commented:
Current issue is that i think i broke the housekeeping loop, adding some debugging to find out why.  A multi-minute house loop should be more obvious.

ehashman reviewed (commented):
I only got through a bit of this but wanted to hit submit before I go OOO for the long weekend.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1494,39 +1497,15 @@ func (kl *Kubelet) Run(updates <-chan kubetypes.PodUpdate) {
 // the most accurate information possible about an error situation to aid debugging.
 // Callers should not throw an event if this operation returns an error.
 func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	klog.Infof("DEBUG: syncPod %s/%s(%s) enter", o.pod.Namespace, o.pod.Name, o.pod.UID)
+	defer klog.Infof("DEBUG: syncPod %s/%s(%s) exit", o.pod.Namespace, o.pod.Name, o.pod.UID)
```
Yeah I'm just leaving this here so i get it on all test runs (not just those with v4 on)

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1494,39 +1497,15 @@ func (kl *Kubelet) Run(updates <-chan kubetypes.PodUpdate) {
 	podStatus := o.podStatus
 	updateType := o.updateType
 
-	// if we want to kill a pod, do it now!
```
Once a pod transitions from "syncing" to "terminating" the pod worker stops calling syncPod and starts calling syncTerminatingPod.  To cause this transition:

1. Delete the pod on apiserver (config change leads to deletion timestamp)
2. Transition restart never pod to succeeded or failed
3. Invoke UpdatePod{UpdateType: Kill} (or use the killPodNow wrapper) for eviction, preemption, and graceful shutdown

This is a one way transition.

Soft admission failures simply call killPod in syncPod, since the next sync might start the container again (soft admission failure is not a termination, it's temporary stop).

smarterclayton replied:
Will be in the next push.

smarterclayton commented:
Note on kind tests I'm seeing 10-20 pauses on container runtime operations, mostly due to a large number of pods being cleaned up in the background (due to the overuse of force deletion).  Some kubelet killPods are taking tens of seconds or longer (when grace period is set to 1 second)

smarterclayton commented:
I'm also seeing flakes in containerd that I can't reproduce on crio.  Places where we make a live status call and then don't end up with the status I know the container runtime has (i.e. the container exits 1, but then status in the kubelet is not up to date).  Will have to look into that to understand what is causing containerd failures.

smarterclayton commented:
Ok, so I think some of the problems with termination were grace period 0 force deletes causing significant build up of pods on the kubelet - when i had HandlePodCleanups doing that synhcronously the housekeeper got overwhelmed, and i think some containerd calls / docker calls were taking much longer than I naively expected (KillPod with grace 1 i expect to take no more than 2s).  I moved that cleanup back to async and then also fixed some other sync things, will add a metric to track so we can find out whether there is a substantial issue or not.

smarterclayton commented:
/retest

BenTheElder commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce-ubuntu-containerd

smarterclayton commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102344/pull-kubernetes-e2e-kind/1398097086230040576 failed because a status update went to the apiserver that didn't contain terminal status.  this *seems* like it was due to an older version of the status getting sent, and then the deletion request was processed.  We may need to move the termination status defaulting (to 137 / empty) from TerminatePod() into the final step of deleting the object, although it may be there is a more subtle race present.

ehashman commented:
/assign

ehashman commented:
/cc @rphillips

BenTheElder commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102344/pull-kubernetes-e2e-gce-ubuntu-containerd/1400210984517242880/ -- probing container startup test should be fixed now ™️ 
https://github.com/kubernetes/kubernetes/pull/101738
/test all

jingxu97 commented:
/assign

smarterclayton commented:
/retest

ehashman commented:
/assign @Random-Liu

Random-Liu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+				}
+				// if this is a retry of termination or a delete shorten, we expect cache to eventually get
+				// filled and can thus simply use the last syncTime
+				status, err = p.podCache.GetNewerThan(update.Pod.UID, lastSyncTime)
```
If we are OK with waiting for the new state to be populated, we should use `GetNewerThan` here, we can even use `time.Now()` as the minTime.

If we don't want to wait for it for some reason, we should call an `UpdateCache` here.

Passing `status = nil` makes it a bit hard to understand the code.

smarterclayton replied:
I was uncomfortable with having podWorker update the status cache - that's sync*Pods job right now (to refresh).  I didn't feel like the logic for syncTerminating/TerminatedPod was more readable by having status refresh outside of their loops.  Agree status == nil was wierd, but I think it might actually possible to get into syncTerminatingPod and status manager has never gotten status.  

We are not ok with waiting for updated state here, because it's possible that you delete a pod BEFORE syncPod gets far enough to create any containers (especially when context cancellation goes in), and therefore you would have to wait for the next pleg refresh (thus increasing the amount of time a create then delete takes).  In general I think we should bias towards lower latency on delete/kill/evict, so I didn't want to wait in those cases.

I considered passing the defaulted Get status, but wanted to distinguish the case where we definitely want to refresh the state of the pod before delete (and syncTerminatingPod felt like the right place to handle it).  Let me mull on this a bit though if that's not sufficiently convincing.

Random-Liu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+
+	if status, ok := p.podSyncStatus[pod.UID]; ok && !status.terminatingAt.IsZero() && !status.terminating {
```
In what case would `status.terminatingAt.IsZero()` not be zero here?

smarterclayton replied:
More as a guard against programmer error, but never (could change it to an assertion)

smarterclayton commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -987,6 +930,21 @@ func (kl *Kubelet) PodResourcesAreReclaimed(pod *v1.Pod, status v1.PodStatus) bo
 			return false
 		}
 	}
+	// // pod's containers should be deleted are allowed to be removed after the pod terminates
```
Just here so you can review the removal, would be removed before final merge.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -104,10 +166,57 @@ const (
+	evicted       bool
+
+	terminatedAt time.Time
+	finished     bool
```
Good point, will add for each.  In this case finished is "all sync work can stop (no more invocations of sync*Pod)

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -286,10 +789,67 @@ func (p *podWorkers) checkForUpdates(uid types.UID) {
+// to UpdatePods for new pods. It returns a map of known workers that are not finished
+// with a value of SyncPodTerminated, SyncPodKill, or SyncPodSync depending on whether
+// the pod is terminated, terminating, or syncing.
+func (p *podWorkers) SyncKnownPods(desiredPods []*v1.Pod) map[types.UID]kubelettypes.SyncPodType {
```
Yeah I was lazy, I'll probably have two types - one for kubelet -> pod worker and one for pod worker internal -> cleanup

smarterclayton replied:
See the review commit pushed to the PR (testing it to make sure I didn't regress anything)

smarterclayton commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1086,206 +1034,114 @@ func (kl *Kubelet) HandlePodCleanups() error {
+			// This termination is aggressive because we don't already have a termination request (i.e.
+			// a deleting static pod should already be terminating). If a static pod is removed from config
+			// and the kubelet restarts, it will be aggressively killed here.
+			if _, ok := allPodsByUID[runningPod.ID]; !ok {
```
Will do.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -40,40 +40,98 @@ import (
+	// CouldHaveRunningContainers returns true before the pod workers have synced,
+	// once the pod workers see the pod (syncPod could be called), and returns false
+	// after the pod has been terminated (running containers guaranteed stopped).
+	CouldHaveRunningContainers(uid types.UID) bool
```
I started with that path, and over repeated reads of the code found it very difficult to understand the loops people were using in that context.  Everyone besides core kubelet loop (volume, gc, etc) should probably be describing in terms of setup and teardown, where setup is racing against the sync pod loop, and teardown is racing against the termination loops. So I wanted to define these in terms of windows:

```
|------------------------------------------| (lifetime of pod object)
     |-------------| (syncing)
|--------------------------------| (acceptable to create)
...--|                           |------------------------------------------------- .... (acceptable to teardown some components)
...--|                                     |------------------------------------------- .... (acceptable to teardown all components)
```

I didn't feel Terminating or Terminated captured those last two accurately, and in fact were confusing (because you could maybe think terminated lasted up until the pod was deleted).

ehashman replied:
> Terminating(), Terminated() and Deleting()

It's unclear to me what would be the difference between Terminating and Deleting, @Random-Liu 

Agreed that it does seem like there's a lot of new conditions defined, possibly more than needed.

@smarterclayton perhaps we should be signaling which are intended for internal use in the kubelet loop vs other consumers by use of public interface vs. private module-specific stuff?

smarterclayton replied:
Terminating means the pod is shutting down.  Deleting means the pod is being deleted.  Those are orthogonal.  A deleted pod can aggressively have content removed by eviction.  A terminating pod has no content being removed.  

* Evicted
* Deleted
* Terminal (succeeded / failed)
* Termination requested
* Termination started
* Terminated
* Gone (orphan)

These are all different states.  Should* hides those conditions into consistent behaviors.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -40,40 +40,98 @@ import (
+	// not be used in cleanup loops because it will return false if the pod has already
+	// been cleaned up - use ShouldPodContainersBeTerminating instead. Also, this method
+	// may return true while containers are still being initialized by the pod worker.
+	IsPodTerminationRequested(uid types.UID) bool
```
What's our most consistent word used?  `Desired`?

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -139,71 +257,207 @@ type podWorkers struct {
+// terminating, or terminated, and will transition to terminating if deleted on the apiserver, it is
+// discovered to have a terminal phase (Succeeded or Failed), or if it is evicted by the kubelet.
+func (p *podWorkers) UpdatePod(options UpdatePodOptions) {
+	// only the pod worker can use the SyncPodTerminated update type
```
Yeah, I was minimizing change (but I think i'll switch to having two different structs, internal and external, and two different update types, internal and external)

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+
+	// never lose an OnCompleteFunc or a PodStatusFunc
+	if undelivered, ok := p.lastUndeliveredWorkUpdate[pod.UID]; ok {
+		if undelivered.OnCompleteFunc != nil {
```
If there are two kills (delete at 30s then evict at 1s), and the second one shortens, we want to terminate the first on complete.  Or at least, that's why I was preserving it.  I think OnCompleteFunc is a misfeature and we should instead accept a channel that we guarantee close at a certain lifecycle point - that would fix killPodNow and also be safer (the use of OnCompleteFunc in UpdatePod is really wrong anyway and sync metrics need to be calculated by the worker looking at podSyncStatus)

smarterclayton replied:
I went back and looked at killPodNow.  We really didn't need OnCompleteFunc and it is complex.  I instead put something more specific in (which is an optional channel on KillPodOptions that is closed when we finally transition from terminating -> terminated).  Also moved sync worker inline.  PodStatusFunc still has to be accumulated, but now only KillPodOptions are accumulated.  PodStatusFunc is accumulative because you could preempt then evict (which I think is correct), but could be more complex.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+	}
+
+	// queue a request if we have decided not to cancel the existing request
+	p.lastUndeliveredWorkUpdate[pod.UID] = options
```
the context cancel PR (which is not here for brevity) will tweak this slightly.

smarterclayton replied:
I added that to the review commit, this changes the ordering here so this isn't a dangling behavior (and you can see where the context cancel is processed)

smarterclayton commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -431,10 +433,40 @@ func (m *manager) updateStatusInternal(pod *v1.Pod, status v1.PodStatus, forceUp
 
 	normalizeStatus(pod, &status)
+
+	// DEBUG ONLY: see what we're getting
```
Yup, will do once we sort out all the others (without this it was almost impossible to debug places where we were racing on status conditions, which after we get through review I'd take out or move to v(6) or something)

smarterclayton replied:
Yes, that's the intention once we get through review.

smarterclayton replied:
Moved to v5.

smarterclayton commented on pkg/volume/util/util.go:
```diff
@@ -381,7 +381,10 @@ func GetUniqueVolumeNameFromSpec(
 // IsPodTerminated checks if pod is terminated
 func IsPodTerminated(pod *v1.Pod, podStatus v1.PodStatus) bool {
-	return podStatus.Phase == v1.PodFailed || podStatus.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(podStatus.ContainerStatuses))
+	// TODO: the guarantees provided by kubelet status are not sufficient to guarantee it's safe to ignore a deleted pod,
```
This is called from controller code so can't access the volume controller.  I was going to file an issue to sig-storage that they may need to ask more state to be returned by kubelet (perhaps a new condition) before they start tearing things down.  Alternatively, they may no longer need this "optimization".

I think the bug right now is that a pod in Succeeded can have 2 containers that haven't started yet reported, then kubelet starts those (so it's not safe to detach), then kubelet catches up and terminates the containers and reports status.  We might actually need a "Terminal" condition for Succeeded/Failed pods if so.

Random-Liu replied:
@jingax10 I remember the issue you mentioned to me is that:
1) Volume manager sees the deletion timestamp and no containers running, so unmount the volume;
2) Pod worker is actually still trying to start new containers;
3) New containers come up with host path mount without the volume.

Does 1) happen in kubelet or the controller? If it happens in kubelet, this change should fix the issue. Or else this change will still not be sufficient.

msau42 replied:
cc correct @jingxu97

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -139,71 +257,207 @@ type podWorkers struct {
+			becameTerminating = true
+		}
+	}
+	if pod.DeletionTimestamp != nil || isRuntimePod {
```
Will push in next

smarterclayton commented:
(waiting to see green before I push the next set of review content)

pohly commented:
> In addition, removing containers now no longer blocks final pod deletion in the API server and are handled as background cleanup.

For the "volume in use" protection controller we need to known when kubelet is done with unmounting volumes for a pod (= NodeUnpublishVolume called for CSI). This used to check for "pod object removed" and got relaxed to "pod object is marked for deletion (= has deletionTimestamp)" to solve a circular dependency with ephemeral PVCs owned by the pod:
https://github.com/kubernetes/kubernetes/blob/1f3ede50f77c41e8d86a6c8cc309d05cb5528342/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L311-L339

Is that still valid?

smarterclayton commented:
> For the "volume in use" protection controller we need to known when kubelet is done with unmounting volumes for a pod (= NodeUnpublishVolume called for CSI). This used to check for "pod object removed" and got relaxed to "pod object is marked for deletion (= has deletionTimestamp)" to solve a circular dependency with ephemeral PVCs owned by the pod:

What is the exact guarantee do you need?  Currently at this time "done unmounting" is not something Kubelet communicates to you, and not only that, but could take arbitrarily long to do (due to this bug and others).  If your guarantee needs to be "kubelet is done unmounting", we'd need to change something.  If your guarantee needs to be "kubelet can start unmounting", deletion, pod succeeded, pod failed are all acceptable.

pohly commented:
> What is the exact guarantee do you need?

What the PVC protection controller currently waits for is "NodeUnpublishVolume was called for the pod(s) using the volume". Once it knows that, it allows the PVC to be deleted, which in turn may end up deleting the volume.

There is already one gap in the current design: deletion is allowed to proceed *before* "NodeUnstageVolume" was called, something that happens after pod removal (https://github.com/kubernetes/kubernetes/issues/94108). So the protection is incomplete, but it's still better than none. The remaining gap needs to be covered by CSI drivers, the only problem is that apparently not all handle this well.

> Currently at this time "done unmounting" is not something Kubelet communicates to you, and not only that, but could take arbitrarily long to do (due to this bug and others).

As it was described to me by @liggitt and @msau42, unmounting was something that kubelet did before deleting the pod. If unmounting failed, it didn't delete the pod. That's why checking the `deletionTimestamp` was seen as giving enough information to determine whether kubelet is done with the pod and its volumes.

 > If your guarantee needs to be "kubelet is done unmounting", we'd need to change something.

Probably.

Isn't this change also affecting users? I'm pretty sure that currently, `NodeUnpublishVolume` failure calls show up as events for the pod, so users can see them with `kubectl describe my-pod`. That no longer works if `NodeUnpublishVolume` is called after removing that pod object.

smarterclayton commented:
> As it was described to me by @liggitt and @msau42, unmounting was something that kubelet did before deleting the pod. If unmounting failed, it didn't delete the pod. That's why checking the deletionTimestamp was seen as giving enough information to determine whether kubelet is done with the pod and its volumes.

DeletionTimestamp being set is *before* the pod is deleted.  The kubelet guarantees that when the pod is actually deleted in etcd (when GET pod/foo returns 404) volumes are unmounted, but not all pods are deleted - for instance, pods that are Succeeded or Failed (terminal phase).  However, today the Kubelet can set Succeeded or Failed before all resources are cleaned up - which is what my comment was.

Today, the Kubelet does not communicate via the API about a pod being "done with volumes" (unmounted volumes) for all pods (only fully deleted pods).  You need to identify whether you need "kubelet will be done with volumes eventually" (deletionTimestamp != nil / phase = Succeeded or Failed) or whether you need the stronger "kubelet is definitely done with volumes" (does not exist today for all pods).  If the latter, we need to add something to the kubelet (perhaps a condition).

msau42 commented:
> You need to identify whether you need "kubelet will be done with volumes eventually" (deletionTimestamp != nil / phase = Succeeded or Failed) or whether you need the stronger "kubelet is definitely done with volumes" (does not exist today for all pods). If the latter, we need to add something to the kubelet (perhaps a condition).

The PVC protection feature needs a signal that "kubelet is done with volumes". Today it relies on kubelet setting the [deletiongraceperiod to 0](https://github.com/kubernetes/kubernetes/blob/5404b5a28a2114299608bab00e4292960dd864a0/pkg/kubelet/status/status_manager.go#L580-L592) when it does the final deletion of the Pod object.

smarterclayton commented:
> The PVC protection feature needs a signal that "kubelet is done with volumes". Today it relies on kubelet setting the deletiongraceperiod to 0 when it does the final deletion of the Pod object.

What happens when you have a restart never pod that reaches completion that uses a volume?  That pod will never be deleted (or rather, can take arbitrarily long to get deleted).

pohly commented:
> What happens when you have a restart never pod that reaches completion (Succeeded / Failed) that uses a volume? That pod will never be deleted (or rather, can take arbitrarily long to get deleted).

Currently the volume deletion will be prevented, which isn't ideal. A new condition that explicitly signals "volumes unmounted" would be a useful improvement.

But my original question was around the current check. If "deletionTimestamp set, gracePeriodSeconds is 0" can now occur before volumes are unmounted, then volume deletion will be allowed to proceed too soon. That would be a regression.

smarterclayton commented:
> But my original question was around the current check. If "deletionTimestamp set, gracePeriodSeconds is 0" can now occur before volumes are unmounted, then volume deletion will be allowed to proceed too soon. That would be a regression.

GracePeriodSeconds can't occur before unmount, so no, no regression here.

smarterclayton commented:
slow looks unrelated, but 

/retest

to be sure

jingxu97 commented on pkg/volume/util/util.go:
```diff
@@ -381,7 +381,10 @@ func GetUniqueVolumeNameFromSpec(
-	return podStatus.Phase == v1.PodFailed || podStatus.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(podStatus.ContainerStatuses))
+	// TODO: the guarantees provided by kubelet status are not sufficient to guarantee it's safe to ignore a deleted pod,
+	// even if everything is notRunning (kubelet does not guarantee that when pod status is waiting that it isn't trying
+	// to start a container).
```
does it mean that even if IsPodTerminated() return true, after that, pod might still try to start a container?

jingxu97 replied:
talked to @Random-Liu about the logic here and checking the use case of isPodTerminated in this volume util, I think we do want to modify the function here so that if isPodTeminated() return true, there will be no more containers trying to start.

smarterclayton replied:
> it could satisfy our use case to modify the function here so that if isPodTeminated() return true, there will be no more containers trying to start.

It is impossible today for components other than the Kubelet (who read the pod api) to know when the kubelet will no longer restart containers, because the Kubelet does not communicate that via pod status in a way that components can use it.  So if you need that, Kubelet needs a change, or you need to weaken your guarantees.

Within the Kubelet, all kubelet sync loops should NOT be using pkg/volume/util (and I have changed code in the kubelet to not use it) and instead use the PodStateProvider style interfaces that the pod worker provides, which *can* guarantee the check.

jingxu97 commented:
is there a function could tell "containers can't be started from now on"?

smarterclayton commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -230,7 +230,7 @@ func (cgc *containerGC) evictContainers(gcPolicy kubecontainer.GCPolicy, allSour
 	if allSourcesReady {
 		for key, unit := range evictUnits {
-			if cgc.podStateProvider.IsPodDeleted(key.uid) || (cgc.podStateProvider.IsPodTerminated(key.uid) && evictTerminatedPods) {
+			if cgc.podStateProvider.ShouldPodContentBeRemoved(key.uid) || (evictTerminatingPods && cgc.podStateProvider.ShouldPodContainersBeTerminating(key.uid)) {
```
ShouldPodContentBeRemoved is "no containers are running and no containers can be started"

smarterclayton replied:
You're right, it is only terminated pods.  This should really be ShouldPodRuntimeBeRemoved, and the name of the variable should be `evictNonDeletedPods` to match the generic interface.  Changing.

smarterclayton commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -782,7 +783,8 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
 			// not a real error.
-			if m.podStateProvider.IsPodDeleted(pod.UID) {
+			// TODO: this is probably not needed now that termination is part of the sync loop
+			if m.podStateProvider.ShouldPodContentBeRemoved(pod.UID) {
```
The bug is more that CRI/CNI/CSI should return typed errors, and we should ignore typed errors of the type "pod not found" when we are confident that those map to "not removable".  

Note that if the pod is marked as deleted due to an observed config change this method does see it - but with context cancellation we would be aborting the sync method before we hit this point (sync methods being interruptible means we are just exiting).  That fixes other problems anyway more straightforwardly.

smarterclayton commented:
> is there a function could tell "containers can't be started from now on"?

It depends on whether you need to know whether the pod has ever existed or not (are you looking at the volumes on disk and trying to determine whether to tear them down, or looking at whether a pod from config should have volumes set up?).  We have coarser checks because people were using the more granular checks incorrectly and inconsistently.  If you can describe the use case better, I can a) point to the method and b) improve the diagram in https://github.com/kubernetes/kubernetes/pull/102344/files#diff-c6d21f5481c391a3f32bcfddf45c9c20132b2876e8be9690289474b76b71d809R269 to cover the question.

ehashman commented on pkg/kubelet/pod_workers.go:
```diff
@@ -286,10 +789,67 @@ func (p *podWorkers) checkForUpdates(uid types.UID) {
+	return workers
+}
+
+func (p *podWorkers) removeWorker(uid types.UID, now time.Time) {
```
AFAICT now isn't used anywhere in here.

smarterclayton replied:
Removed now, renamed, added godoc.

ehashman reviewed (commented):
I've read through this PR now and mostly understand the gist. I think it would have been better to write this as a series of smaller, more targeted changes that we could merge individually to reduce risk of regressions, particularly given how low unit test coverage in the kubelet is.

As is, I think it's probably best if we merge this ASAP and let it bake in CI as long as possible. I think this does address a serious issue with the kubelet's pod lifecycle that's causing all sorts of subtle race condition bugs. Unfortunately, I'm not sure if, as written, this gets all of them. For that sort of structural assurance, ideally I wish that we could send pods through a DAG, but I'm not sure that's feasible at this point and this PR does take us closer to where we want to go.

smarterclayton commented:
> AFAICT now isn't used anywhere in here.

removeWorker is called by SyncKnownPods

smarterclayton commented on test/e2e/apps/statefulset.go:
```diff
@@ -756,6 +756,10 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ginkgo.By("Waiting until pod " + podName + " will start running in namespace " + f.Namespace.Name)
+			if err := e2epod.WaitForPodNameRunningInNamespace(f.ClientSet, podName, f.Namespace.Name); err != nil {
+				framework.Failf("Pod %v did not start running: %v", podName, err)
+			}
```
The test is trying to verify that statefulPodName goes into failed at least once, and that the stateful controller sees that and recreates it, and that the trigger is the scheduling of a conflicting port.  If a human introduces a bug in the stateful set that causes it to never start, or a bug in the kubelet that fails pods incorrectly, or another random failure happens, it might not be the conflicting port (admission) that causes it to get recreated.  By waiting for the first pod to start successfully, we move those failures to earlier (we are more confident that the conflicting port is the cause).  This move means debugging is easier (if the pod fails to start it's more likely a bug in the test, or a bug in kubelet startup, if the pod starts but then conflict doesn't cause a restart, it's because someone broke admission in the kubelet)

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1562,16 +1536,15 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 	runnable := kl.canRunPod(pod)
 	if !runnable.Admit {
-		// Pod is not runnable; update the Pod and Container statuses to why.
+		// Pod is not runnable; and update the Pod and Container statuses to why.
+		if apiPodStatus.Phase != v1.PodFailed && apiPodStatus.Phase != v1.PodSucceeded {
+			apiPodStatus.Phase = v1.PodPending
+		}
```
It does not duplicate that?  The check condition is "if the pod is in a phase other than a terminal phase, move the phase to pending".  The check you describe as duplicate is "if the phase is terminal, block moving back to a non-terminal".

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1699,8 +1686,9 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
-	if !kl.podIsTerminated(pod) {
+	// Volume manager will not mount volumes for terminating pods
+	// TODO: once context cancellation is added this check can be removed
+	if !kl.podWorkers.IsPodTerminationRequested(pod.UID) {
```
We are in a sync* method here, which means we are running in the pod worker and we know for certain the pod is known to the kubelet.  This method is checking whether in the time since syncPod has started, has someone asked us to stop.  That is the equivalent to the previous check 

> for someone running in the pod worker, should this pod keep attempting to create containers?

The `ShouldPodContainersBeTerminating` method is different.  It is running *outside* a pod worker, and can't return true until `syncPod` exits.  So the syncPod gets to know early (because it's the one creating containers) and all other workers have to wait until syncPod exits.

This is the core race the PR addresses - that other loops need to know FOR SURE that no new containers can be started, which only pod worker knows.  The pod worker itself needs to know when syncPod has been requested to stop early, but syncPod at this line of code also knows that containers could still get started in the next loop.  In the future, when context cancellation happens, instead of continuing on to sync pod, we would exit earlier from syncPod, which means that the volume manager sees ShouldPodContainersBeTerminating earlier.  At that point we would remove IsPodTerminationRequested.

I can change this to simply be an exit early, or I can have it check for context cancellation, but that's bring more scope in.  The correct version of this code looks like:

```
func syncPod(ctx context.Context) { 
   ...
   if err := kl.volumeManager.WaitForAttachAndMount(ctx,` pod); err != nil {
     if err == context.Canceled {
       return err
     }
     ... // existing debug logic
  }
  result := kl.containerRuntime.SyncPod(ctx, ...)
  if result is cancelled {
    return context.Canceled
  }
  ... // err checking
  return nil
}
```

There is no deadlock possible that I'm aware of because SyncPod will fail when mounts aren't ready at container create or container start (which is why we wait here in normal paths).

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1731,6 +1719,155 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	// to call syncTerminatingPod until container status is recalculated. Instead, the first
+	// invocation of syncTerminatingPod on a given pod may send a nil podStatus to allow us
+	// to more quickly guarantee shutdown of everything the last syncPod created.
+	if podStatus == nil {
```
You're right.  That update also needs to be conditional on timestamp (I think) - since updateCache() is not done under lock.  Really, only PLEG should be able to invoke Set(), and PLEG needs to either rely on cache to enforce timestamp ordering, or PLEG itself needs to enforce timestamp ordering (but since cache has the lock and the data, it's probably cache for SetIfNewer()).

smarterclayton replied:
See https://github.com/kubernetes/kubernetes/pull/102344/commits/416b068dab13bc7710c1cce9b5f3072bf93ae21d which i think is the correct way because only PLEG can control happens-before on status read.

Are we guaranteed to have the updated cache by the time the next syncTerminating is called?  I don't think we do

249043822 replied:
Agree that we should not directly update pod cache here

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -1731,6 +1719,155 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	sort.Strings(containers)
+	klog.V(4).InfoS("DEBUG: State post termination", "pod", klog.KObj(pod), "podUID", pod.UID, "containers", strings.Join(containers, " "))
+	// TODO: seen on containerd in kind runs
+	if len(runningContainers) > 0 {
```
This is definitely an assertion - if someone calls KillPod with stale cache then you can get here still have containers running, so I expect the assertion to remain and the logging to be at V(4) or above.

Random-Liu commented:
Sorry for the delay, will take a second round review this week. :)

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -2106,9 +2219,9 @@ func (kl *Kubelet) HandlePodAdditions(pods []*v1.Pod) {
 		}
 
-		if !kl.podIsTerminated(pod) {
+		if !kl.podWorkers.IsPodTerminationRequested(pod.UID) {
```
agreed, will add a comment

smarterclayton commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1086,206 +1034,114 @@ func (kl *Kubelet) HandlePodCleanups() error {
 	// running pods to clean up the volumes.
 	// TODO: Evaluate the performance impact of bypassing the runtime cache.
-	runningPods, err = kl.containerRuntime.GetPods(false)
+	runningRuntimePods, err = kl.containerRuntime.GetPods(false)
```
In this case it's allowing the pod worker to own all termination logic.  Effectively this is going to create a worker for "orphan pods" and the orphan pod worker then unifies all the handling (so orphaned volume, lagging controller, etc) is set up.  So this is kind of the reverse of SyncKnownPods - we *could* simply pass this in to pod worker and have that method handle it, but I'd prefer to do the loop here in a predictable order.  Is that a better way of framing why this makes sense to do?

249043822 replied:
Currently podworker maintains podsyncstatus,  why this line directly calling runtime still needed? I think it should be replaced like `(dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods()`

smarterclayton commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -311,7 +311,7 @@ func (cgc *containerGC) evictSandboxes(evictTerminatedPods bool) error {
 
 	for podUID, sandboxes := range sandboxesByPod {
-		if cgc.podStateProvider.IsPodDeleted(podUID) || (cgc.podStateProvider.IsPodTerminated(podUID) && evictTerminatedPods) {
+		if cgc.podStateProvider.ShouldPodContentBeRemoved(podUID) || (evictTerminatingPods && cgc.podStateProvider.ShouldPodContainersBeTerminating(podUID)) {
```
Same

ehashman commented:
> > AFAICT now isn't used anywhere in here.
> 
> removeWorker is called by SyncKnownPods

I meant the `now` parameter, was replying on @Random-Liu's thread

rphillips commented:
@smarterclayton: I think this PR is getting close. @harche has tested it against a Bugzilla in our queue and it does fix the issue. Are we to the point of moving the debug log messages to a different log level?

smarterclayton commented:
The log messages have been updated.

smarterclayton commented:
The current issue is:

1. Decision on how to mutate the cache status (probably with an UpdateCache method on PLEG) based on a @Random-Liu comment and my response
2. Remove the last debug statements once that is in place

Random-Liu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -139,77 +364,233 @@ type podWorkers struct {
+	default:
+		workType = SyncPodWork
+
+		// completely ignore KillPodOptions
```
When could this happen? Should we return an error instead?

smarterclayton replied:
We could simply have another public method besides UpdatePod which is "kill pod"

Random-Liu replied:
Let's at least add comment to point out that this is not a valid case.

Random-Liu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,76 +602,412 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+		if originalKillOpts := undelivered.Options.KillPodOptions; originalKillOpts != nil {
+			// once a pod is killed, all subsequent UpdatePod methods must be "kill" and so we should
+			// be able to carry the PodStatusFunc over
+			if originalKillOpts.PodStatusFunc != nil {
```
Do we need to accumulate `PodStatusFunc`? There is only one pod status, and later one will overwrite the previous one? In this case, can we just keep the latest one?

I'm still not quite comfortable with adding this recursive function if not needed.

If the caller wants to perform operations with side effect in the function, that is a misusage, they should probably use the `CompleteCh` instead in that case.

smarterclayton replied:
So in theory you could evict a pod and then the user could delete it faster (so you'd lose the eviction state) by reducing the grace period.  That's the primary bug this fixes.

You could also delete a pod and then have it evicted, in which case we need both data to pass through.  You could also be preempted then evicted - what status should be on the pod?

Random-Liu replied:
Most of the functions are just updating `Phase`, `Message` and `Reason`, and the updates are applied to the `status` passed in this function, not to the status manager or apiserver, so the change history is discarded directly.

In this case, what status has gone through doesn't actually matter, only the latest status matters.
* If newer status doesn't completely overwrite the old status, the generated mixed status may be confusing anyway;
* If newer status does completely overwrite the old status, the old status is not useful.

This seems only to be meaningful to me if we update the status manager in between.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -40,40 +40,98 @@ import (
+	// been cleaned up - use ShouldPodContainersBeTerminating instead. Also, this method
+	// may return true while containers are still being initialized by the pod worker.
+	IsPodTerminationRequested(uid types.UID) bool
+	// ShouldPodContainersBeTerminating returns true before the pod workers have synced,
```
Yeah, reworded.

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -104,27 +209,147 @@ const (
+}
+
+func (s *podSyncStatus) IsWorking() bool            { return s.working }
+func (s *podSyncStatus) IsTerminating() bool        { return !s.terminatingAt.IsZero() }
```
Fair.

smarterclayton replied:
Renamed to `startedTerminating / HasStartedTerminating` and `IsTerminationRequested`

249043822 replied:
Maybe `IsTerminationRequested `and `IsTerminationStarted`

smarterclayton commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,76 +602,412 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+					// cache at all as it may be out of date and we want to begin termination as quickly
+					// as possible
+					lastKillTime = time.Now()
+					status = nil
```
So I debated that - but the use case for terminating is

a) i need to ensure i have the latest for killPod - i prefer to have that code close to the call (not buried in the pod worker) so that it's clear that you need the latest state for kill pod
b) after we kill the pod we MUST know that all containers are stopped - a crio bug is fatal and should be caught, and we ALSO want to ensure the latency is low between kill (when no containers can run) and when we refresh the cache (so that if containers are purged, we have the status up to date so we don't lose the container status)

So for those reasons I prefer to keep the status refresh close to the kill, even though I think it's cleaner as you are describing.

Random-Liu replied:
The new commit LGTM.

ehashman commented:
```
--- FAIL: TestSyncTerminatingPodKillPod (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x2f44565]
```

looks like this needs a fix @smarterclayton

249043822 commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,76 +602,412 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+	}
+
+	klog.V(4).InfoS("Pod has been terminated and is no longer known to the kubelet, remove all history", "podUID", uid)
+	delete(p.podSyncStatuses, uid)
```
For static pod,  pod.uid does not change, while pod REMOVE and ADD encountes race condition,  `delete(p.podSyncStatuses, uid)` here may cause deadlock,  as the old podwoker may write workUpdate to new podworker's podUpdate channel  in `completeWorkQueueNext(uid types.UID) `

smarterclayton replied:
Thanks, we don't have a test case for this in node tests, which I'll add.

smarterclayton replied:
Turns out we have bugs in this area - a pod created like this doesn't seem to have its mirror pod updated.  Digging into why that happens before I finish up the area here.

smarterclayton replied:
https://github.com/kubernetes/kubernetes/pull/103374 is a test that fails in master (indicating the mirror pod not getting updated).

Random-Liu replied:
> For static pod, pod.uid does not change, while pod REMOVE and ADD encountes race condition, delete(p.podSyncStatuses, uid) here may cause deadlock, as the old podwoker may write workUpdate to new podworker's podUpdate channel in completeWorkQueueNext(uid types.UID)

Can you elaborate more about how the deadlock happens? IIUC, `SyncKnownPods` runs `HandlePodCleanups`, and during which no desired pod will be added/removed. In that case, if the static pod is seen, we keep the pod worker; if it is not seen, we remove the pod worker.

Why would `delete(p.podSyncStatuses, uid)` cause a deadlock?

249043822 replied:
> > For static pod, pod.uid does not change, while pod REMOVE and ADD encountes race condition, delete(p.podSyncStatuses, uid) here may cause deadlock, as the old podwoker may write workUpdate to new podworker's podUpdate channel in completeWorkQueueNext(uid types.UID)
> 
> Can you elaborate more about how the deadlock happens? IIUC, `SyncKnownPods` runs `HandlePodCleanups`, and during which no desired pod will be added/removed. In that case, if the static pod is seen, we keep the pod worker; if it is not seen, we remove the pod worker.
> 
> Why would `delete(p.podSyncStatuses, uid)` cause a deadlock?
> 
> But if a static is moved out of the manifest directory and then moved back, it is not quite clear to me what would happen. I feel like the termination will continue because the uid (generated from pod hash) is not changed.

@Random-Liu The case mybe like this below(We have encountered in v1.19 by add `delete(p.isWorking, uid)` here): 
1. For a static pod A terminated, `SyncKnownPods` calls `removeTerminatedWorker`, but at this time, the old podworker does not absolutely exited, it may have not run to the line `p.completeWorkQueueNext(pod.UID)`
2. Just this time. the static pod A manifest comes back, as  `removeTerminatedWorker` clears all the data like  `p.podUpdates[uid]`, so UpdatePod() will create a new podwoker. and the condtion below still satisfies, then the now podworker becomes running
```
if !status.IsWorking() {
		status.working = true
		podUpdates <- work
		return
	}
```
 so at this mini time, there are two podworders runing with some same shared data struct like `p.podUpdates[uid]` and `p.lastUndeliveredWorkUpdate[pod.UID]`
3. At this time, there are still some podUpdates come, UpdatePod() will put the podUpdate to  `p.lastUndeliveredWorkUpdate[pod.UID]`, just then , the old podworker run into the func `p.completeWorkQueueNext(pod.UID)`, and it writes a podUpdate to p.podUpdates[uid] channel and exits 
4. Just this time,  UpdatePod() will still put a podUpdate to  `p.lastUndeliveredWorkUpdate[pod.UID]`, just then , the new podworker run into the func `p.completeWorkQueueNext(pod.UID)`, as the p.podUpdates[uid] channel can only put one data. so it could not write successfully, then deadlock causes

Random-Liu replied:
For 1.

The current code only actually `removeTerminatedWorker` when the pod worker `finished`. And the pod worker only `finished` at the end of `completeTerminated`, and the pod worker will return right after that.

I don't see how the case you described could happen.

249043822 replied:
@Random-Liu Thanks for your ack.  I'm sorry for that I ignored the `managePodLoop  `will exit after `completeTerminated`, I mentioned the case frontend we have found in original code. Now this new change don't have this nit, I have no problem,

Random-Liu commented on pkg/kubelet/kubelet.go:
```diff
@@ -1734,6 +1726,143 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	// see the status of terminated containers (since we could race with other cleanup processes on the
+	// node in low resource situations).
+	// TODO: have KillPod return the terminal status of stopped containers
+	podStatus, err := kl.pleg.WaitForUpdatedStatus(&kubecontainer.Pod{ID: pod.UID, Namespace: pod.Namespace, Name: pod.Name})
```
Offline discussed.

Only this call to the container runtime to get pod status is required for correctness.

The other direct container runtime pod status call and pod cache update are mostly for optimization: reducing latency. We probably want to solve the race condition and ensure correctness first.

The latency reduction can be done in a separate change, and we can evaluate different options, e.g.:
1) Make `GetPodStatus` lightweight, so that we are more open to the breakglass call for latency sensitive scenario to bypass 1s PLEG relist;
2) Make PLEG event driven, so that status can be updated (no control flow critical) faster.

Those may need more thoughts, and can be a separate change. :)

smarterclayton replied:
Pushed review comment 1 with updates to match our discussion.

Random-Liu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -40,58 +42,161 @@ import (
+	// use ShouldPod*().
+	IsPodTerminationRequested(uid types.UID) bool
+
+	// ShouldPodContainersBeTerminating returns true once pod workers have synced,
```
nit: returns false before pod workers have synced.

Random-Liu commented on pkg/kubelet/pod_workers.go:
```diff
@@ -221,76 +602,396 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+		}()
+
+		switch {
+		case err == context.Canceled:
```
I understand what we are trying to do. However, the sync function doesn't use context today, having this log may be confusing if the grace period is really shortened.

Can we remove the context related change and graceful period shorten logic from this PR for now? It doesn't seem to take any effect right now, and can be added in a following up change.

smarterclayton replied:
I really don't want to remove the grace period logic shortening in the prep step, since if we get an error in syncTerminatingPod and have to retry I think we should be using the shortened period.  Note that cancellation changes the core structure, so if we're going to soak I want to soak with the context in place.  I agree it isn't used, so I guess we could simplify some parts, but then we'd bring those back without the infra I expect.

I agree that having it in without testing doesn't accomplish much - I'd prefer to spend the time now adding the unit tests then for verifying context cancel requeues in the right scenarios even if no one is cancelling, vs waiting to add them since it's so fundamental to the idea that we *can* sync.  Once we start cancelling, it'll be pretty obvious where the bugs are (especially with e2es on pods with long preStop hooks), so the overall cancellation stuff benefits from having a fairly well tested core loop with one touch point.

Random-Liu commented:
Status: The only thing that is not finalized about this PR is "how to handle `PodStatusFunc` from different kill pod attempt".

If the grace period shorten logic in this change actually being used without context support? If not, can we punt that to the next change, and get the race conditions fixed first? With that, we can get this change in earlier, and punt the `PodStatusFunc` overwritten discussion.

smarterclayton commented:
> If the grace period shorten logic in this change actually being used without context support? If not, can we punt that to the next change, and get the race conditions fixed first? With that, we can get this change in earlier, and punt the PodStatusFunc overwritten discussion.

Since it impacts the structure of pod workers (how we queue them) I'd prefer to have the right mechanism in place to preserve  It is definitely a bug - multiple kill sources overwrite each other arbitrarily (randomly) depending on whether an error occurs during syncPod / during killPod - so I don't like leaving the behavior defined in some form, but I'm ok with saying "we'll only honor the first or the last pod status write".  It's just that we have to explicitly decide one or the other.

The following scenario needs to be deterministic:

1. Pod is RestartNever
2. Container 1 exits with failure and a status message (thus the pod should be transitioning to Failed)
3. An eviction is triggered on the pod trying to set failed and reason to Evicted

Do we want the pod to have the status from 2 or 3?

smarterclayton commented:
> If the grace period shorten logic in this change actually being used without context support?

Grace period shorten is used if syncTerminatingPod returns an error (due to a transient failure in killPod).  So I'd prefer to keep the shortened interval the next time we invoke killPod, or go one step further and invoke killPod the second time with the shorter interval.

Note however that that is yet another bug - we don't have a declarative way of telling CRI to send the kill signal as distinct from wait.  So if we have to killPod a second time (which we will in practice because CRI is a remote call) we're sending two TERM signals (which is wrong, although we don't guarantee we only send one) and then we're waiting up to twice as long for the kill.  Instead, we should be sending TERM to CRI, then waiting in kubelet, then sending KILL at the interval.  I didn't try to fix that bug here, but it needs to be fixed for people using long grace periods and I plan to do that going forward.

smarterclayton commented:
/retest

Random-Liu commented:
> Grace period shorten is used if syncTerminatingPod returns an error (due to a transient failure in killPod).

Hm, I might have missed something. When `syncTerminatingPod` returns an error, it will requeue and retry later. However, I don't see we calculate the grace period based on time passed?

> Instead, we should be sending TERM to CRI, then waiting in kubelet, then sending KILL at the interval.

That behavior is currently implemented by the container runtime. Kubelet sends `StopContainer` with a timeout, and expects the container runtime to send `SIGTERM` to the container, and `SIGKILL` after the timeout.

> Note that cancellation changes the core structure, so if we're going to soak I want to soak with the context in place. I agree it isn't used, so I guess we could simplify some parts, but then we'd bring those back without the infra I expect.

Just thought we could simplify the change and make it easier and safer to get in. If we want to keep that part, I'm also OK with that. The new `statusPostTerminating` LGTM.

Random-Liu commented:
/lgtm
/approve

Random-Liu commented:
/hold

Will double check with @dchen1107 about whether we should get this change in before code freeze. Will unhold once we get a confirmation.

ehashman commented:
> On one hand, we do want this change, which fixes long standing issues, and opens new possibilities for the future; on the other hand, this change refactors the core pod lifecycle management logic, which may need longer soaking time.

FWIW, given the impact by fixing a number of long-standing issues, and the length of freeze for 1.22 (a full four weeks), I'd be inclined to take this now rather than wait.

mrunalp commented:
:+1: to taking this in now.

smarterclayton commented:
> Hm, I might have missed something. When syncTerminatingPod returns an error, it will requeue and retry later. However, I don't see we calculate the grace period based on time passed?

No, but we do pass the shorter grace period the second time.  For very long grace periods (hours), if someone does send a shorter grace period and the call happens to fail (for any reason), the second grace period being faster is a nice net win.  I agree the real change we need is to calculate the desired "KILL" time based on when term is sent, then send term and wait up to that (subsequent shortenings need only send KILL at the shorter interval), but that's further out.

> That behavior is currently implemented by the container runtime. Kubelet sends StopContainer with a timeout, and expects the container runtime to send SIGTERM to the container, and SIGKILL after the timeout.

Right, I missed this during CRI implementation and didn't catch it in review.  We shouldn't have done that, and instead designed CRI to send signals that kubelet manages (TERM then KILL) given the design goals of pod termination.  That's mostly on me - I didn't spend enough time on it.

dims commented:
@Random-Liu i am inclined to get this in as well. there's time to revert if needed. will remove hold later today unless we get a clear -1 from @dchen1107 by then.

dims commented:
/milestone v1.22

dchen1107 commented:
Had offline discussions and review with @Random-Liu. We all agreed the design is very clear and there is no alternative way to get this in easily. Also agreed with @ehashman and @Random-Liu above,  4 weeks soaking time is long enough especially with the improved testing coverage in this PR. 

/approve
/unhold

smarterclayton commented:
/retest

smarterclayton commented:
Integration failures appear to be unrelated, but will continue to monitor if they recur here.

smarterclayton commented:
As an unrelated note (this PR does not impact either way), the original trigger for this race detection was #102025 and the observation that when our e2e tests were force deleting most pods the depth of kubelet backlog was being hidden.  After rebasing that on this, some tests fail because it takes over a minute from the time the CRI finishes cleanup for the status of the deleted pod to be sent back to the apiserver (status updates from kubelet are serialized and under load can take a very long time).  When we close the "default grace period to zero on a pod" hole (vs simply doing it at delete time) we'll start seeing status latency impact test flakes - which will give us good reason to focus on status latency over the smaller and less impactful sync latency we discussed here.

ehashman commented:
flake #103512 
/retest

BenTheElder commented:
#103512 again
/retest
